lib/http3/server.c

```diff
@@ -1500,8 +1505,71 @@ h2o_http3_conn_t *h2o_http3_server_accept(h2o_http3_server_ctx_t *ctx, quicly_ad
     return &conn->h3;
 }
 
+static void graceful_shutdown_close_stragglers(h2o_timer_t *entry)
+{
+    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http3._graceful_shutdown_timeout, entry);
+    h2o_linklist_t *node, *next;
+
+    /* We've sent two GOAWAY frames, close the remaining connections */
+    for (node = ctx->http3._conns.next; node != &ctx->http3._conns; node = node->next) {
+        struct st_h2o_http3_server_conn_t *conn = H2O_STRUCT_FROM_MEMBER(struct st_h2o_http3_server_conn_t, _conns, node);
+        next = node->next;
+        h2o_quic_close_connection(&conn->h3.super, 0, "shutting down");
+    }
+}
+
+static void graceful_shutdown_resend_goaway(h2o_timer_t *entry)
+{
+    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http3._graceful_shutdown_timeout, entry);
+    h2o_linklist_t *node;
+    int do_close_stragglers = 0;
+
+    /* HTTP/3 draft section 5.2.8 --
+     * "After allowing time for any in-flight requests or pushes to arrive, the endpoint can send another GOAWAY frame
+     * indicating which requests or pushes it might accept before the end of the connection.
+     * This ensures that a connection can be cleanly shut down without losing requests. */
+    for (node = ctx->http3._conns.next; node != &ctx->http3._conns; node = node->next) {
+        struct st_h2o_http3_server_conn_t *conn = H2O_STRUCT_FROM_MEMBER(struct st_h2o_http3_server_conn_t, _conns, node);
+        if (conn->h3.state < H2O_HTTP3_CONN_STATE_HALF_CLOSED && quicly_get_state(conn->h3.super.quic) == QUICLY_STATE_CONNECTED) {
+            quicly_stream_id_t max_stream_id = quicly_get_remote_next_stream_id(conn->h3.super.quic, 0 /* == bidi */) - 4;
+            h2o_http3_send_goaway_frame(&conn->h3, max_stream_id);
+            conn->h3.state = H2O_HTTP3_CONN_STATE_HALF_CLOSED;
+            do_close_stragglers = 1;
+        }
+    }
+
+    /* After waiting a second, we still had active connections. If configured, wait one
+     * final timeout before closing the connections */
+    if (do_close_stragglers && ctx->globalconf->http3.graceful_shutdown_timeout) {
+        h2o_timer_unlink(&ctx->http3._graceful_shutdown_timeout);
```

Comment:
This line is unnecessary. By the timer invokes its callback, it is guaranteed to have been unlinked.

Reply:
I see. Actually the same code exists in http2. Can we remove it from http2 as well?

Reply:
Removed from http2 as well. We can easily revert it if I was not supposed to.

Reply:
Ah thanks. I'll fix the http2 side separately.